File: examples/mpi/c/putfence.c
/* %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
   %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
   %%%%
   %%%% This program file is part of the book and course
   %%%% "Parallel Computing for Science and Engineering"
   %%%% by Victor Eijkhout, copyright 2013-8
   %%%%
   %%%% putfence.c : point-to-point through put & fence
   %%%%
   %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
   %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
*/

#include <stdlib.h>
#include <mpi.h>
#include <stdio.h>
#include <unistd.h>

int main(int argc,char **argv) {

#include "globalinit.c"

  //codesnippet putfence
  MPI_Win the_window;
  //codesnippet end
  int my_number=0, window_data[2], other = nprocs-1;
  if (procno==0)
    my_number = 37;

  //codesnippet putfence
  MPI_Win_create
    (&window_data,2*sizeof(int),sizeof(int),
     MPI_INFO_NULL,comm,&the_window);
  MPI_Win_fence(0,the_window);
  if (procno==0) {
    MPI_Put
      ( /* data on origin: */   &my_number, 1,MPI_INT,
        /* data on target: */   other,1,    1,MPI_INT,
        the_window);
  }
  MPI_Win_fence(0,the_window);
  //codesnippet end
  if (procno==other)
    printf("I got the following: %d\n",window_data[1]);
  //codesnippet putfence
  MPI_Win_free(&the_window);
  //codesnippet end

  MPI_Finalize();
  return 0;
}

File: examples/mpi/p/putfence.py
#!/usr/bin/env python
################################################################
################################################################
####
#### This program file is part of the book and course
#### "Parallel Computing"
#### by Victor Eijkhout, copyright 2013-6
####
#### putfence.py : point-to-point through put & fence
####
################################################################
################################################################

import numpy as np
import random # random.randint(1,N), random.random()
import sys

from mpi4py import MPI

comm = MPI.COMM_WORLD
procid = comm.Get_rank()
nprocs = comm.Get_size()
if nprocs<2:
    print("C'mon, get real....")
    sys.exit(1)

#snippet windispp
intsize = np.dtype('int').itemsize
window_data = np.zeros(2,dtype=int)
win = MPI.Win.Create(window_data,intsize,comm=comm)
#snippet end

#snippet putblockp
my_number = np.empty(1,dtype=int)
src = 0; tgt = nprocs-1
if procid==src:
    my_number[0] = 37
else:
    my_number[0] = 1
win.Fence()
if procid==src:
    # put data in the second element of the window
    win.Put(my_number,tgt,target=1)
win.Fence()
#snippet end

if procid==tgt:
    print("Window after put:",window_data)


